packages/clay-data-provider/src/useResource.ts

```diff
@@ -0,0 +1,236 @@
+
+	let retryDelayIntervalId = useRef<null | NodeJS.Timeout>(null).current;
+
+	const firstRenderRef = useRef<boolean>(true);
+
+	const cache = useCache(
+		fetchPolicy,
+		storage,
+		storageMaxSize,
+		link,
+		variables
+	);
+
+	const debouncedVariablesChange = useDebounce(variables, fetchDelay);
+
+	const setNetworkStatus = (status: NetworkStatus) => {
+		networkStatus = status;
+		onNetworkStatusChange(status);
+	};
+
+	const handleRefetch = () => {
+		doFetch();
+		setNetworkStatus(NetworkStatus.Refetch);
+	};
+
+	const cleanRetry = () => {
+		if (retryDelayIntervalId) {
+			clearInterval(retryDelayIntervalId);
+		}
+	};
+
+	const getRetryDelay = (retryAttempts: number) => {
+		const {delay = {}} = fetchRetry;
+		const {initial = 300, jitter = true} = delay;
+
+		const baseDelay = jitter ? initial : initial / 2;
+
+		let value = Math.min(baseDelay * 2 ** retryAttempts);
+
+		if (jitter) {
+			value = Math.random() * value;
+		}
+
+		return value;
+	};
+
+	const handleFetchRetry = (err: any, retryAttempts: number) => {
+		const {attempts = 5} = fetchRetry;
+
+		cleanRetry();
+
+		if (attempts > 0 && retryAttempts < attempts) {
+			const delay = getRetryDelay(retryAttempts);
+
+			// eslint-disable-next-line no-console
+			console.warn(
+				`DataProvider: Trying ${retryAttempts +
+					1} of ${attempts} will happen in ${delay}ms`
+			);
+
+			retryDelayIntervalId = setInterval(() => {
+				doFetch(retryAttempts + 1);
+			}, delay);
+		} else {
+			setNetworkStatus(NetworkStatus.Error);
+			// eslint-disable-next-line no-console
+			console.error('DataProvider: Error making the requisition', err);
+		}
+	};
+
+	const cleanPoll = () => {
+		if (pollingIntervalId) {
+			clearInterval(pollingIntervalId);
+		}
+	};
+
+	const fetchOnComplete = (result: any) => {
+		// Should clear retry interval if any of the
+		// attempts are successful.
+		cleanRetry();
+
+		setResource(result);
+		setNetworkStatus(NetworkStatus.Unused);
+
+		cache.set(result);
+
+		if (pollInterval > 0) {
+			cleanPoll();
+
+			pollingIntervalId = setInterval(() => {
+				maybeFetch(NetworkStatus.Polling);
+			}, pollInterval);
+		}
+	};
+
+	const getUrlFormat = (
+		link: string,
+		variables: TVariables,
+		fetchOptions?: RequestInit
+	) => {
+		if (fetchOptions && fetchOptions.method !== 'GET') {
+			return link;
+		}
+
+		const query = getQueryParams(variables);
+
+		return query ? `${link}/${query}` : link;
+	};
+
+	const doFetch = (retryAttempts = 0) => {
+		let promise: Promise<any>;
+
+		switch (typeof link) {
+			case 'function':
+				promise = link();
+				break;
+			case 'string':
+				promise = fetch(
+					getUrlFormat(link, variables, fetchOptions),
+					fetchOptions
+				).then(res => res.json());
+				break;
+			default:
+				return null;
+		}
+
+		timeout(fetchTimeout, promise)
+			.then(fetchOnComplete)
+			.catch(err => handleFetchRetry(err, retryAttempts));
+	};
+
+	const maybeFetch = (status: NetworkStatus) => {
+		const cacheData = cache.get();
```

Comment:
@wincent this may seem a little strange because it is not going to `key` here, but I ended up implicitly leaving the key in `useCache` because we can have the cache disabled and I did not want to have to use `stringify` at all times unnecessarily that a variables change.